pkg/trace/sampler/coresampler.go

```diff
@@ -102,40 +107,141 @@ func (s *Sampler) Start() {
 	}()
 }
 
-// update decays scores and rate computation coefficients
-func (s *Sampler) update() {
-	s.Backend.DecayScore()
-	s.AdjustScoring()
+// countWeightedSig counts a trace sampled by the sampler and update rates
+// if buckets are rotated
+func (s *Sampler) countWeightedSig(now time.Time, signature Signature, n float32) bool {
+	bucketID := now.Unix() / int64(bucketDuration.Seconds())
+	s.muSeen.Lock()
+	prevBucketID := s.lastBucketID
+	s.lastBucketID = bucketID
+
+	// pass through each bucket, zero expired ones and adjust sampling rates
+	updateRates := prevBucketID != bucketID
+	if updateRates {
+		s.updateRates(prevBucketID, bucketID)
+	}
+
+	buckets, ok := s.seen[signature]
+	if !ok {
+		buckets = [numBuckets]float32{}
+	}
+	buckets[bucketID%numBuckets] += n
+	s.seen[signature] = buckets
+	s.muSeen.Unlock()
+	atomic.AddInt64(&s.totalSeen, 1)
+	return updateRates
 }
 
-func (s *Sampler) report() {
-	kept, seen := s.Backend.report()
-	metrics.Count("datadog.trace_agent.sampler.kept", kept, s.tags, 1)
-	metrics.Count("datadog.trace_agent.sampler.seen", seen, s.tags, 1)
+// updateRates distributes TPS uniformly on each signature and apply it to the moving
+// max of seen buckets.
+// Rates increase are bounded by 20% increases, it requires 13 evaluations (1.2**13 = 10.6)
+// to increase a sampling rate by 10 fold in about 1min.
+func (s *Sampler) updateRates(previousBucket, newBucket int64) {
+	if len(s.seen) == 0 {
+		return
+	}
+	tpsPerSig := s.targetTPS.Load() / float64(len(s.seen))
+
+	s.muRates.Lock()
+	defer s.muRates.Unlock()
+
+	rates := make(map[Signature]float64, len(s.seen))
+	for sig, buckets := range s.seen {
+		maxBucket, buckets := zeroAndGetMax(buckets, previousBucket, newBucket)
+		s.seen[sig] = buckets
+
+		seenTPS := float64(maxBucket) / bucketDuration.Seconds()
+		rate := 1.0
+		if tpsPerSig < seenTPS && seenTPS > 0 {
+			rate = tpsPerSig / seenTPS
```

Comment:
nit: Put a comment saying: seenTPS is the total number of traces generated in the tracer (vs the total number of traces sent to the agent). So since we want to generate tpsPerSig P1, we need to put the rate of P1s at number of P1s we want / total number of traces.